src/analyses/base.ml

```diff
@@ -1997,6 +1997,20 @@ struct
     let st' = invalidate ~deep:false ~ctx (Analyses.ask_of_ctx ctx) gs st shallow_addrs in
     invalidate ~deep:true ~ctx (Analyses.ask_of_ctx ctx) gs st' deep_addrs
 
+  let check_free_of_non_heap_mem ctx special_fn ptr =
+    match ctx.ask (Queries.MayPointTo ptr) with
+    | a when not (Queries.LS.is_top a) ->
+      let warn_if_not_heap_var special_fn var =
+        if not (ctx.ask (Queries.IsHeapVar var)) then
+          M.warn ~category:(Behavior (Undefined InvalidMemoryDeallocation)) ~tags:[CWE 590] "Free of non-dynamically allocated memory in function %s for pointer %a" special_fn.vname d_exp ptr
+      in
+      let pointed_to_vars =
+        Queries.LS.elements a
+        |> List.map fst
+      in
+      List.iter (warn_if_not_heap_var special_fn) pointed_to_vars
+    | _ -> ()
```

Comment:
When `a`  is top, we should probably also warn!

Reply:
Makes sense, yes! Thanks for pointing out! Should be covered in https://github.com/goblint/analyzer/pull/1099/commits/acb1ca83d00ae5f09167e1b30b775a465736ecd1